### PR TITLE
feat(lint-unicorn): add rule prefer set has

### PR DESCRIPTION
**Pull request by jordan-boyer**
implementing the rule [prefer set has](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-set-has.md) from unicorn.

I put the fix as dangerous fix because they was no test for the fix on the unicorn part. I did add some test for this but i'm not sure that i'm covering everything.

### CODE REVIEW

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 1]
```diff
@@ -0,0 +1,900 @@
...
+    "splice",
+    "toReversed",
+    "toSorted",
+    "toSpliced",
+    "with",
+};
+
+fn prefer_set_has_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Switch to `Set`")
+        .with_help("should be a `Set`, and use .has()` to check existence or non-existence.")
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:34]:
```suggestion
        .with_help("should be a `Set`, and use `.has()` to check existence or non-existence.")

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 2]
```diff
@@ -0,0 +1,900 @@
...
+    if member_expr.is_computed() || member_expr.optional() {
+        return false;
+    }
+
+    let Some(callee_name) = member_expr.static_property_name() else {
+        return false;
+    };
+
+    callee_name == "includes"
+}
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:82]: use `is_method_call` ? 

https://github.com/oxc-project/oxc/blob/1e2f012d3c795ca9015c287ecf1768bd4ee2dd72/crates/oxc_linter/src/ast_util.rs#L316

**Review by camc314 — COMMENTED** [review]
this looks great

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 1 above)
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:34]: These messages are reversed: the `warn()` string should describe the error/issue, the help text should explain how to fix it:

```suggestion
    OxcDiagnostic::warn("Should be a `Set`, and use `.has()` to check existence or non-existence.")
        .with_help("Switch to `Set`")
```

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 3]
```diff
@@ -0,0 +1,900 @@
...
+        _ => false,
+    }
+}
+
+fn is_multiple_calls(node: &AstNode, ctx: &LintContext, root_scope_id: ScopeId) -> bool {
+    let mut id = node.id();
+    let mut was_in_root_scope = node.scope_id() == root_scope_id;
+    loop {
+        let Some(parent) = ctx.nodes().parent_node(id) else { return false };
+        let parent_scope = parent.scope_id();
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:120]: Could we refactor this loop to use `ctx.nodes().iter_parents()` instead? Something like this:


```suggestion
    for parent in ctx.nodes.iter_parents(node.id()).skip(1) {
        let parent_scope = parent.scope_id();
```

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+                        let span = Span::new(start, end);
+                        declaration_fix.push(fixer.replace(span, "])"));
+                    } else {
+                        declaration_fix.push(fixer.insert_text_before(&call_expr.span, "new Set("));
+                        declaration_fix.push(fixer.insert_text_after(&call_expr.span, ")"));
+                    }
+                }
+                Expression::NewExpression(new_expr) => {
+                    let start = new_expr.span.start + 4;
+                    let end = start + 6;
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:250]: Could we replace these number constants with something computable at compile time, or at least a named variable?

Something like this:

```suggestion
                    const set_start = "Set(["
                    const set_start_len = set_start.len();
                    let start = new_expr.span.start + set_start_len;
                    let end = start + 6;
```

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 5]
```diff
@@ -0,0 +1,900 @@
...
+            let node = ctx.nodes().get_node(reference.node_id());
+            let Some(parent) = ctx.nodes().parent_node(node.id()) else {
+                return true;
+            };
+            let Some(parent_of_parent) = ctx.nodes().parent_node(parent.id()) else {
+                return true;
+            };
+            let AstKind::CallExpression(call_expression) = parent_of_parent.kind() else {
+                return true;
+            };
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:209]: can be replaced with `parent_kind`:

```suggestion
            let Some(AstKind::CallExpression(call_expression)) = ctx.nodes().parent_kind(parent.id()) else {
                return true;
            };
```

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 6]
```diff
@@ -0,0 +1,900 @@
...
+            let mut references_fix = fixer.new_fix_with_capacity(len);
+            let references = symbol_table.get_resolved_references(symbol_id);
+            for reference in references {
+                let node = ctx.nodes().get_node(reference.node_id());
+                let Some(parent) = ctx.nodes().parent_node(node.id()) else {
+                    continue;
+                };
+                let AstKind::MemberExpression(member_expr) = parent.kind() else {
+                    continue;
+                };
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:270]: I don't think we need the actual node here, we just need its ID? We can also use `parent_kind` directly:

```suggestion
                let Some(AstKind::MemberExpression(member_expr)) = ctx.nodes().parent_node(reference.node_id()) else {
                    continue;
                };
```

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 7]
```diff
@@ -0,0 +1,900 @@
...
+    let mut was_in_root_scope = node.scope_id() == root_scope_id;
+    loop {
+        let Some(parent) = ctx.nodes().parent_node(id) else { return false };
+        let parent_scope = parent.scope_id();
+        if was_in_root_scope && parent_scope != root_scope_id {
+            return false;
+        }
+        was_in_root_scope = parent_scope == root_scope_id;
+        let parent_kind = parent.kind();
+        if matches!(
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:126]: would it break anything to move this check up earlier? checking `AstKind` is faster than getting the parent node, so doing fewer calls to `parent_node` should be faster.

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs [hunk 8]
```diff
@@ -0,0 +1,900 @@
...
+    callee_name == "includes"
+}
+
+fn is_array_of_or_from(name: &str, callee: &MemberExpression) -> bool {
+    if name != "from" && name != "of" {
+        return false;
+    }
+
+    return callee.object().get_identifier_reference().map_or(false, |ident| ident.name == "Array");
+}
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:90]: would `is_specific_member_access` work here?

```suggestion
fn is_array_of_or_from(callee: &MemberExpression) -> bool {
    callee.is_specific_member_access("Array", "of") || callee.is_specific_member_access("Array", "from")
}
```

**Review by camchenry — APPROVED** [review]
looks pretty good, just had some suggestions on a few things, but overall I think this rule makes sense to me!

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 5 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:209]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 6 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:270]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 8 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:90]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 1 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:34]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 2 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 7 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:111]: Here i'm looping over all the parent and only existing the loop if we exist the scope of the variable declaration or if i found a match for multiple call. So i'm not sure if i can do any change here

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 3 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:105]: done

File: crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs (hunk 4 above)
Comment (jordan-boyer) [reply crates/oxc_linter/src/rules/unicorn/prefer_set_has.rs:244]: done

**Review by camchenry — APPROVED** [review]

**Review by camc314 — APPROVED** [review]
thank you!

**Pull request merged**